src/irmin-test/node.ml

```diff
@@ -4,48 +4,131 @@ let check pos typ ~expected actual =
-    let k1 = random_key () and k2 = random_key () in
-    let a =
-      X.empty () |> with_binding "a" (`Node k1) |> with_binding "b" (`Node k2)
-    in
-    check __POS__ [%typ: int] ~expected:2 (X.length a)
+    let with_binding k v t = Map.add t k v in
+    let d1 = Map.random_data () and d2 = Map.random_data () in
+    let a = Map.empty () |> with_binding "1" d1 |> with_binding "2" d2 in
+    check __POS__ [%typ: int] ~expected:2 (Map.length a)
 
   let test_remove () =
     (* Remove is a no-op on an empty node *)
-    check __POS__ [%typ: X.t] ~expected:(X.empty ()) X.(remove (empty ()) "foo")
+    check __POS__ [%typ: Map.t] ~expected:(Map.empty ())
+      Map.(remove (empty ()) "foo")
+
+  let test_find () =
+    let bindings = random_bindings 256 in
+    let node = map_of_bindings bindings in
+    bindings
+    |> List.iter (fun (k, v) ->
+           check __POS__ [%typ: Map.data option] ~expected:(Some v)
+             (Map.find node k))
+
+  let test_equal () =
+    let module Map = struct
+      include Map
+
+      type nonrec t = t [@@deriving irmin ~equal ~to_bin_string ~of_bin_string]
+    end in
+    let bindings = random_bindings 256 in
+    let m = map_of_bindings bindings in
+
+    let m_rev = map_of_bindings (List.rev bindings) in
+    check __POS__ [%typ: bool] ~expected:true (Map.equal m m_rev);
+
+    let m_subset = map_of_bindings (List.tl bindings) in
+    check __POS__ [%typ: bool] ~expected:false (Map.equal m m_subset);
+
+    let m_serialised =
+      m |> Map.to_bin_string |> Map.of_bin_string |> Result.get_ok
+    in
+    check __POS__ [%typ: bool] ~expected:true (Map.equal m m_serialised)
 
   let suite =
     [
-      Alcotest.test_case "empty" `Quick test_empty;
-      Alcotest.test_case "add" `Quick test_add;
-      Alcotest.test_case "remove" `Quick test_remove;
+      ("empty", test_empty);
+      ("add", test_add);
+      ("remove", test_remove);
+      ("find", test_find);
+      ("equal", test_equal);
     ]
 end
+
+module Make (Make_node : Irmin.Node.Generic_key.Maker) : sig
+  val suite : unit Alcotest.test_case list
+end = struct
+  (* For each [Node] maker, we can instantiate the test suite above twice: once
+     for regular nodes, and once for portable nodes. *)
+
+  module Schema = Irmin.Schema.KV (Irmin.Contents.String)
+  module Hash = Schema.Hash
+  module Key = Irmin.Key.Of_hash (Hash)
+  module Node = Make_node (Hash) (Schema.Path) (Schema.Metadata) (Key) (Key)
+
+  type key = Key.t [@@deriving irmin]
+
+  module Extras = struct
+    type data = [ `Node of Key.t | `Contents of Key.t * unit ]
+    [@@deriving irmin]
+
+    let random_data =
+      let hash_of_string = Irmin.Type.(unstage (of_bin_string Hash.t)) in
+      let random_string =
+        Irmin.Type.(unstage (random (string_of (`Fixed Hash.hash_size))))
+      in
+      fun () ->
+        match hash_of_string (random_string ()) with
+        | Error _ -> assert false
+        | Ok x -> (
+            match Random.int 2 with
```

Comment:
Are these randoms seeded deterministically?

Reply:
I was about to say "no", but it looks like we don't call `Random.self_init` in `test/irmin/test.ml` at the moment, which means we're just getting the default state.

I've added an extra commit that does the seeding to be consistent with tests that use `Irmin_test.run`.

Reply:
Thanks